flink-libraries/flink-table/src/main/java/org/apache/flink/table/runtime/batch/io/RFCCsvInputFormat.java

```diff
@@ -0,0 +1,449 @@
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		return endOfSplit;
+	}
+
+	private Object[] projectedFields(Object[] record) throws IOException {
+
+		Object[] resultantRecord = new Object[fieldTypes.length];
+		int index = 0;
+		for (int i = 0; i < this.fieldIncluded.length; i++) {
+
+			try {
+				if (fieldIncluded[i]) {
+					resultantRecord[index++] = record[i];
+				}
+			}
+			catch (Exception e) {
+				throw new IOException();
+			}
+		}
+		return resultantRecord;
+	}
+
+	private long findFirstDelimiterPosition() throws IOException{
+
+		this.stream.seek(this.getSplitStart());
+		if (this.stream.getPos() == 0) {
+			return 0;
+		}
+		else {
+			int pos = 1;
+			while ((this.stream.read()) != this.recordDelimiter.charAt(0)) {
+				pos++;
+			}
+			return pos;
+		}
+	}
+
+	private long findLastDelimiterPosition() throws IOException{
+
+		this.stream.seek(this.splitStart + this.splitLength);
+		int pos = 0;
+		char c;
+		int read = this.stream.read();
+		while (read != -1) {
+			c = (char) read;
+			if (c == this.recordDelimiter.charAt(0)) {
+				break;
+			}
+			else {
+				read = this.stream.read();
+			}
+			pos++;
+		}
+		return pos;
+	}
+
+	private Object[] castRecord(Object[] record)  throws IOException {
+
+		if (isLenient()) {
+			for (int i = 0; i < this.fieldTypes.length; i++) {
+				try {
+					record[i] = dataTypeConversion(record[i], fieldTypes[i]);
+				} catch (Exception e) {
+					throw new IOException(e);
+				}
+			}
+		}
+		else {
+			for (int i = 0; i < this.fieldTypes.length; i++) {
+				try {
+					record[i] = dataTypeConversion(record[i], fieldTypes[i]);
+				} catch (Exception e) {
+					throw new IOException(e);
+				}
+			}
+		}
+		return record;
+	}
+
+	public <I, O> O dataTypeConversion(I input, Class<O> outputClass) throws Exception {
+
+		String type = outputClass.getSimpleName().toLowerCase();
+		try {
+			switch (type) {
+				case "string":
+					return (O) input.toString();
+				case "boolean":
+					if (input.toString().length() != 0) { // special case for boolean
+						return (O) Boolean.valueOf(input.toString());
+					}
+					return null;
+				case "byte":
+					return (O) Byte.valueOf(input.toString());
+				case "short":
+					return (O) Short.valueOf(input.toString());
+				case "integer":
+					return (O) Integer.valueOf(input.toString());
+				case "long":
+					return (O) Long.valueOf(input.toString());
+				case "float":
+					return (O) Float.valueOf(input.toString());
+				case "double":
+					return (O) Double.valueOf(input.toString());
+				case "decimal":
+					return (O) Double.valueOf(input.toString());
+				case "date":
+					return (O) Date.valueOf(input.toString());
+				case "time":
+					return (O) Time.valueOf(input.toString());
+				case "timestamp":
+					return (O) Timestamp.valueOf(input.toString());
+				default:
+					return (O) input;
+			}
+		}
+		catch (Exception e) {
+			if (isLenient()) {
+				return null;
```

Comment:
This lenient case is not correctly forwarded. It cannot be distinguished from the case where a field is actually null